pyfarm/agent/utility.py

```diff
@@ -275,3 +273,127 @@ def total_seconds(td):
+                data = loaded_file.read().strip()
+
+            return UUID(data)
+
+        except (OSError, IOError) as e:
+            if e.errno != ENOENT:
+                raise
+
+        except ValueError:
+            cls.log.warning("%r does not contain valid data for a UUID", path)
+            return None
+
+    @classmethod
+    def _save(cls, agent_uuid, path):
+        """
+        Saves ``agent_uuid`` to ``path``.  This classmethod will
+        also create the necessary parent directories and handle
+        conversion from the input type :class:`uuid.UUID`.
+        """
+        assert isinstance(agent_uuid, UUID)
+        assert isinstance(path, STRING_TYPES)
+
+        parent_dir = dirname(path)
+        if parent_dir.strip():
+            try:
+                os.makedirs(parent_dir)
+
+            except (OSError, IOError) as e:
+                if e.errno != EEXIST:
+                    cls.log.warning("Failed to create %s, %s", parent_dir, e)
+                    raise
+
+        try:
+            with open(path, "w") as output:
+                output.write(str(agent_uuid))
+
+            cls.log.debug("Cached %s to %r", agent_uuid, path)
+            return path
+
+        except (OSError, IOError) as e:
+            cls.log.warning("Failed to write %s, %s", path, e)
+            raise
+
+    @classmethod
+    def load(cls, path=None):
+        """
+        This classmethod will attempt to load the agent's UUID from
+        disk.  If it cannot be located ``None`` will be returned.
+
+        :param string path:
+            An explicit path to load the UUID from.  In all other cases
+            we'll attempt to load the UUID from any directory
+            configuration files could load from.
+        """
+        cls.log.debug("Attempting to load agent UUID from disk")
+
+        if path is not None:
+            return cls._load(path)
+
+        for directory in config.directories(validate=False):
+            path = join(directory, "uuid.dat")
+            agent_uuid = cls._load(path)
+            if agent_uuid is not None:
+                cls.log.debug(
+                    "Cached UUID %s was loaded from %r", agent_uuid, path)
+                return agent_uuid
+
+        cls.log.warning(
+            "Unable to locate cached agent UUID, one will be generated.")
+
+    @classmethod
+    def save(cls, agent_uuid, path=None):
+        """
+        This classmethod will save ``agent_uuid`` to ``path`` or
+        to the first available path in the configuration.
+        """
+        assert isinstance(agent_uuid, UUID), agent_uuid
+        cls.log.debug("Attempting to save UUID %s to disk.", agent_uuid)
+
+        if path is not None:
+            return cls._save(agent_uuid, path=path)
+
+        directories = config.directories(validate=False, unversioned_only=True)
+        for directory in directories:
+            path = cls._save(agent_uuid, join(directory, "uuid.dat"))
```

Comment:
Yeah agreed....very bad place to have that stored by default.  This should be fixed in d85aa8d387a474cfb4b4d1dd41b952f12a693a52 which also improved the error handling.